parsing documentation - mark's review (#936)
urbit/docs: tutorials/arvo/parsing.md

## Functional parsers

How parsers are built varies substantially depending on what sort of programming
language it is written in. As Hoon is a functional programming language, we will
be focused on understanding _functional parsers_, also known as _combinator
parsers_. In this section we will make light use of pseudocode, as introducing
the Hoon to describe functional parsers here creates a chicken-and-egg problem.

Complex functional parsers are built piece by piece from simpler parsers
that are plugged into one another in various ways to perform the desired task.

The basic building blocks, or primitives, are parsers that read only a
single character. There are frequently a few types of possible input characters,
such as letters, numbers, and symbols. For example, `parse("1", integer)` calls
the parsing routine on the string `"1"` and looks for an integer, and so it
returns the integer `1`. However, taking into account what was said above about
parsers returning the unparsed portion of the string as well, we should
represent this return as a tuple. So we should expect something like this:
```
> parse("1", integer)
(1, "")
> parse("123", integer)

> (1, "23")

Reviewer: I suggested changing the pseudo-code to look more like hoon (ie `(parse "1" integer)`, `[1 ""]` etc). You gave the comment a :+1:. Is the idea here that you don't want to get it confused for something you might try running?

Author: I thought you were just saying that I should mention that it is pseudocode. I think it definitely makes sense to change the pseudocode to be more Hoon-like though.

be focused on understanding _functional parsers_, also known as _combinator
parsers_. In this section we will make light use of pseudocode, as introducing
the Hoon to describe functional parsers here creates a chicken-and-egg problem.

Complex functional parsers are built piece by piece from simpler parsers
that are plugged into one another in various ways to perform the desired task.

The basic building blocks, or primitives, are parsers that read only a
single character. There are frequently a few types of possible input characters,
such as letters, numbers, and symbols. For example, `(parse "1" integer)` calls
the parsing routine on the string `"1"` and looks for an integer, and so it
returns the integer `1`. However, taking into account what was said above about
parsers returning the unparsed portion of the string as well, we should
represent this return as a tuple. So we should expect something like this:
```
> (parse "1" integer)
[1 ""]
> (parse "123" integer)
[1 "23"]
```
What if we wish to parse the rest of the string? We would need to apply the
`parse` function again:
```
> parse(parse("123", integer), integer)
(12, "3")

> > parse(parse(parse("123", integer), integer), integer)

Reviewer: These two don't match the rest.

[p=[p=1 q=2] q=[~ [p='a' q=[p=[p=1 q=2] q="bc"]]]]
```
We note that `p.edg` is `[p=1 q=2]`, indicating that the next character to be
parsed is in line 1, column 2. `q.edg` is not null, indicating that parsing
succeeded. `p.q.edg` is `'a'`, which is the result of the parse. `p.q.q.edg` is the same as `p.edg`, which is always the case for
`rule`s built using standard library functions when parsing succeeds. Lastly,
`q.q.edg` is `"bc"`, which is the part of the input `tape` that has yet to be parsed.

Now let's see what happens when parsing fails.
```
> =edg ((just 'b') [[1 1] "abc"])
> edg
[p=[p=1 q=1] q=~]
```
Now we have that `p.edg` is the same as the input `hair`, `[1 1]`, meaning the
parser has not advanced since parsing failed. `q.edg` is null, indicating that
parsing has failed.

Later we will use [+star](#star) to string together a sequence of `+just`s in
order to parse multiple characters at once. However, it should be noted that
`+jest`, which is covered next, ought to be used in favor of `+just`, which is

> mostly included here for pedagogical reasons.

Reviewer: As per discussion on the comment in the previous PR, not sure if the disclaimer here is warranted.

Author: Yeah that makes sense. I added it in supposing that you had knowledge that `jest` was actually deprecated regardless of what the comment said.

> `+cold` is a `rule` builder that takes in a constant noun we'll call `cus` and a
> `rule` we'll call `sef` and returns a `rule` identical to the `sef` except it

Reviewer: ```suggestion
`+cold` is a `rule` builder that takes in a constant noun we'll call `cus` and a
`rule` we'll call `sef`. It returns a `rule` identical to the `sef` except it
```

Maybe this applies elsewhere also.

> These functions take in either a `tape` or an `@` (typically representing a
> `cord`) with a `rule` and attempts to parse the input with the `rule`. If the

Reviewer: ```suggestion
These functions take in either a `tape` or a `cord`,
alongside a `rule`, and attempt to parse the input with the `rule`. If the
```

I know they technically take a `@`, but doing parsing on atoms that aren't `@t` (or "deeper", `@tas` etc) is unlikely to be sane.

accepts a `rule`. `+knee` produces a `rule` that implements any recursive calls
in the `rule` in a manner acceptable to the compiler. Thus the preferred manner
to write the above `rule` is as follows:

```hoon

> ++  pars
>   |-
>   ;~  plug  prn
>     ;~  pose  knee  *tape
>       |.  ^$  (easy ~)

Reviewer: ```suggestion
++  pars
  |-
  ;~  plug  prn
    ;~  pose
      (knee *tape |.(^$))
      (easy ~)
    ==
  ==
```

btw, I think it's more common to wrap `+knee` around the parser on the outside (ie, directly under the `|-` here, and as we also do below). This has the helpful effect of specifying a parsing result type early on. I don't think it technically matters much though.

it. If you'd like to follow along, save the following as `expr-parse.hoon` in
your `gen/` folder.
```hoon

> :: expr-parse: parse arithmetic expressions

Reviewer: ```suggestion
::  expr-parse: parse arithmetic expressions
```

::
|=  xprs=tape
|^  (scan xprs tape)
++  factor

>   %+  knee  @ud

Reviewer: ```suggestion
  %+  knee  *@ud
```

This goes for all of these instances.

```hoon
++  factor
  %+  knee  @ud

>   |.  ~+  ;~  pose

Reviewer: ```suggestion
  |.  ~+
  ;~  pose
```

to match the above.

Since expressions ultimately reduce to factors, we are actually building a
recursive rule. Thus we need to make use of `+knee`. The first argument for

> `+knee` is `@ud`, since our final answer should be a `@ud`.

Reviewer: ```suggestion
`+knee` is `*@ud`, since our final answer should be a `@ud`.
```
